dataset/dataset.cpp

```diff
@@ -573,104 +475,216 @@ void DataArrayView::setUnit(const units::Unit unit) const {
+    const auto tmp =
+        is_view_of_unaligned
+            ? makeViewItems(view.parentDims(), view.get_data().coords)
+            : makeViewItems(view.get_data().coords);
+    items.insert(tmp.begin(), tmp.end());
   }
-  if constexpr (std::is_same_v<MapView, CoordsView>) {
-    // Access disabled with nullptr since views of dataset items or slices of
-    // data arrays may not set or erase coords.
-    return MapView({unaligned_ptr ? m_mutableDataset : nullptr, unaligned_ptr},
-                   std::move(items), slices());
-  } else {
-    // Note: Unlike for CoordAccess this is *not* unconditionally disabled with
-    // nullptr since it sets/erase attributes/masks of the *item*.
-    return MapView(
-        {slices().empty() ? m_mutableDataset : nullptr, &name(), unaligned_ptr},
-        std::move(items), slices());
+  deep_ptr<CoordAccess> unaligned_ptr{nullptr};
+  if (is_view_of_unaligned) {
+    // This is a view of the unaligned content of a realigned data array.
+    T unaligned = view.get_data().unaligned->data;
+    const auto tmp =
+        make_coords(unaligned, category, category == CoordCategory::Unaligned);
+    if constexpr (std::is_same_v<T, DataArrayView>)
+      unaligned_ptr = std::make_unique<CoordAccess>(tmp.access());
+    items.insert(tmp.items().begin(), tmp.items().end());
   }
+  if constexpr (std::is_same_v<T, DataArrayView>)
+    return CoordsView(
+        // Coord insert/erase disabled if:
+        // - coords of a slice:
+        //   array['x', 7].coords['x'] = x # fails
+        //   array.coords['x'] = x # ok
+        // - (aligned) coords of a dataset item:
+        //   del ds['a'].coords['x'] # fails
+        //   del ds.coords['x'] # ok
+        // Note that del array.coords['x'] works even if 'x' is unaligned
+        CoordAccess{view.slices().empty() ? &view.get_dataset() : nullptr,
+                    &view.name(), std::move(unaligned_ptr), is_item},
+        std::move(items), view.slices());
+  else
+    return CoordsConstView(std::move(items), view.slices());
 }
 
 /// Return a const view to all coordinates of the data view.
 CoordsConstView DataArrayConstView::coords() const noexcept {
-  return makeView<CoordsConstView>();
+  return make_coords(*this, CoordCategory::All);
 }
 
-/// Return a const view to all attributes of the data view.
-AttrsConstView DataArrayConstView::attrs() const noexcept {
-  return makeView<AttrsConstView>();
+/// Return a view to all coordinates of the data view.
+CoordsView DataArrayView::coords() const noexcept {
+  // Typically view of item of data array, therefore:
+  // ds['a'].coords['x'] = x # inserts unaligned coord
+  return make_coords(*this, CoordCategory::All);
 }
 
+/// Return a const view to all coordinates of the data array.
+CoordsConstView DataArray::coords() const { return get().coords(); }
+
+/// Return a view to all coordinates of the data array.
+CoordsView DataArray::coords() {
+  // Note difference to DataArrayView::coords():
```

Comment:
It is consistent across `Dataset` and `DataArray`, but thus, as you say, slightly inconsistent between view and non-view. I think it is the only way it can be, since the former consistency is more critical, I believe.